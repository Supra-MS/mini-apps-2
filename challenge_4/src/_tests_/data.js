module.exports = {
  winBoard: [
    [
      { mine: true, mineCount: 0, reveal: false, flagged: true },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false }
    ],
    [
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false }
    ],
    [
      { mine: true, mineCount: 0, reveal: false, flagged: true },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false }
    ],
    [
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false }
    ],
    [
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: true, mineCount: 0, reveal: false, flagged: true },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false }
    ],
    [
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: true, mineCount: 0, reveal: false, flagged: true },
      { mine: true, mineCount: 0, reveal: false, flagged: true },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false }
    ],
    [
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: true, mineCount: 0, reveal: false, flagged: true },
      { mine: true, mineCount: 0, reveal: false, flagged: true },
      { mine: true, mineCount: 0, reveal: false, flagged: true },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: true, mineCount: 0, reveal: false, flagged: true }
    ],
    [
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: false, mineCount: 0, reveal: true, flagged: false },
      { mine: true, mineCount: 0, reveal: false, flagged: true }
    ]
  ],

};
