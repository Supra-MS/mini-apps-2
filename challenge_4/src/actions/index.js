const setGameLevel = {
  beginner: {
    boardWidth: 8,
    boardHeight: 8,
    totalMines: 10
  },
  intermediate: {
    boardWidth: 16,
    boardHeight: 16,
    totalMines: 40
  },
  expert: {
    boardWidth: 30,
    boardHeight: 16,
    totalMines: 99
  }
};

const START = 'START';
const TOGGLE_FLAG = 'TOGGLE_FLAG';
const GAME_LEVEL = 'GAME_LEVEL';
const GAME_OVER = 'GAME_OVER';
const WON = 'WON';

const start = () => {
  return {
    type: START
  };
};

const toggle = (flagged) => {
  return {
    type: TOGGLE_FLAG,
    payload: {
      flagged
    },
  };
};

const gameLevel = (level) => {
  return {
    type: GAME_LEVEL,
    payload: {
      level
    },
  };
};

const gameOver = () => {
  return {
    type: GAME_OVER
  };
};

const won = () => {
  return {
    type: WON
  };
};

module.exports = { setGameLevel, START, TOGGLE_FLAG, GAME_LEVEL, GAME_OVER, WON, start, toggle, gameLevel, gameOver, won };