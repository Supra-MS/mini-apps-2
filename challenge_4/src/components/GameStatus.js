import React from 'react';

const GameStatus =({ gameOver, seconds, won, mine, reset}) => {
  return (
    <div className="container">
      <div className="game-status">
        <button className="time-counter">⏱ {!gameOver ? `${seconds}` : `${seconds}`} </button>
        <button onClick={reset} className="reset"> {gameOver ? '😣' : won ? '😎' : '🙂'} </button>
        <button className="mine-counter">💣 {mine} </button>
      </div>
      <div className={`win-status ${won && 'won'} ${gameOver && 'lose'}`}>
        {won ? '🥳 CONGRATS! YOU WON 🥳' : gameOver ? '!! TRY AGAIN !!' : null}
      </div>
    </div>
  );
};

export default GameStatus;