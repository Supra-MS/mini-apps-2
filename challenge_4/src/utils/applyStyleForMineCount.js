export const applyStyleForMineCount = (content, style) => {
  let colors = {
    'ðŸ’£': 'rgba(255, 0, 0, 0.5)',
    1: 'blue',
    2: 'green',
    3: 'red',
    4: 'rgba(234, 0, 255, 0.988)',
    5: '#fff600',
    6: 'deepskyblue',
    7: 'rgb(125, 50, 247)',
    8: '#9162ece0',
  };

  if (content === 'ðŸ’£') {
    style = Object.assign({}, style, { backgroundColor: colors[content] });
    return style;
  } else {
    for (let i = 1; i < 9; i++) {
      if (content === i) {
        style = Object.assign({}, style, { color: colors[content] });
        return style;
      }
    }
  }

};
