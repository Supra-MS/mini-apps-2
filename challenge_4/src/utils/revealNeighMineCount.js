export const revealNeighMineCount = (board, boardWidth, boardHeight, x, y) => {
  let mineCount = 0;
  for (let i = x - 1; i <= x + 1; i++) {
    for (let j = y - 1; j <= y + 1; j++) {
      if ((i < 0 || i >= boardWidth) || (j < 0 || j >= boardHeight) || (i === x && j === y)) {
        continue;
      }
      if (board[i][j].mine) {
        mineCount++;
      }
    }
  }
  return mineCount;
};