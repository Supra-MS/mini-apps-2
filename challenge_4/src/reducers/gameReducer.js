import { setGameLevel, START, TOGGLE_FLAG, GAME_LEVEL, GAME_OVER, WON} from '../actions';

const setInitialState = {
  level: 'beginner',
  mine: setGameLevel['beginner'].totalMines,
  gameOver: false,
  won: false,
};

const gameReducer = (state = setInitialState, action) => {
  switch(action.type) {
    case START: {
      return {
        gameOver: false,
        won: false,
        mine: setGameLevel[state.level].totalMines,
        level: state.level
      };
    }
    case TOGGLE_FLAG: {
      const { flagged } = action.payload;
      let { mine } = state;
      flagged ? mine -= 1 : mine += 1;
      return Object.assign({}, state, { mine });
    }
    case GAME_LEVEL: {
      const { level } = action.payload;
      return {
        gameOver: false,
        won: false,
        mine: setGameLevel[level].totalMines,
        level
      };
    }
    case GAME_OVER: {
      return Object.assign({}, state, { gameOver: true });
    }
    case WON: {
      return Object.assign({}, state, { won: true });
    }
    default: return state;
  }
};

export default gameReducer;
