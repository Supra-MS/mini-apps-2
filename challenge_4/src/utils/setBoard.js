import { setGameLevel } from '../actions';

export const createBoard = (level) => {
  const { boardWidth, boardHeight } = setGameLevel[level];
  const board = Array.from(new Array(boardWidth),
    () => new Array(boardHeight).fill(
      {
        mine: false,
        mineCount: 0,
        reveal: false,
        flagged: false
      }
    )
  );
  return board;
};

const generateRandom = (duplicate, widthOrHeight) => {
  let randomWidth = Math.floor(Math.random() * widthOrHeight);
  return (randomWidth === duplicate) ? generateRandom(duplicate, widthOrHeight) : randomWidth;
};

export const plantMines = (level, firstClick, xFirst, yFirst) => {
  const minePositions = [];
  const { boardWidth, boardHeight, totalMines } = setGameLevel[level];

  while (minePositions.length < totalMines) {
    let x = generateRandom(xFirst, boardWidth);
    let y = generateRandom(yFirst, boardHeight);

    if (minePositions.length) {
      const duplicateMines = minePositions.filter((position) => {
        return (position.x === x && position.y === y);
      });

      let getIndexOfDuplicatedMine = (position) => (position.x === x && position.y === y);
      const isDuplicated = duplicateMines.length > 0;

      if (isDuplicated) {
        minePositions.splice(minePositions.findIndex(getIndexOfDuplicatedMine), 1);
      }

      if (!isDuplicated || firstClick) {
        minePositions.push({ x, y });
      }

    } else {
      minePositions.push({ x, y });
    }

  }
  console.log('Mine Positions: ', minePositions);
  return minePositions;
};

export const setBoard = (level, firstClick, x, y) => {
  const minePositions = plantMines(level, firstClick, x, y)
  const board = createBoard(level);

  for (let position of minePositions) {
    board[position.x][position.y] = Object.assign({}, board[position.x][position.y], { mine: true })
  }
  console.log('Board ðŸ˜‡ â˜Ž ðŸ‡¨ðŸ‡¦', board)
  return { board, minePositions };
};