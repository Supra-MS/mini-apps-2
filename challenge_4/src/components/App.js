import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setGameLevel, toggle, start, gameLevel, gameOver, won } from '../actions';
import { setBoard, createBoard } from '../utils/setBoard';
import { revealNeighMineCount } from '../utils/revealNeighMineCount';
import Board from './Board';
import Title from './Title';
import GameLevel from './GameLevel';
import GameStatus from './GameStatus';

class App extends Component {
  constructor(props) {
    super(props);
    const { level } = this.props;
    this.state = {
      board: createBoard(level),
      seconds: 0,
      firstClick: true,
      minePositions: null,
    };

    this.tick = this.tick.bind(this);
    this.reset = this.reset.bind(this);
    this.changeGameLevel = this.changeGameLevel.bind(this);
    this.handleClickCell = this.handleClickCell.bind(this);
    this.handleContextClickCell = this.handleContextClickCell.bind(this);
  }

  tick() {
    this.setState({
      seconds: this.state.seconds + 1,
    });
  }

  reset(e) {
    e.preventDefault();
    const { level } = this.props;
    this.props.dispatch(start());
    this.setState({
      board: createBoard(level),
      firstClick: true,
      seconds: 0
    });
    clearInterval(this.interval);
  }

  handleClickCell(x, y) {
    const { gameOver, won, level } = this.props;
    const { firstClick } = this.state;
    if (gameOver || won) {
      return;
    }
    if (firstClick) {
      const { board, minePositions } = setBoard(level, firstClick, x, y);
      board[x][y] = Object.assign({}, board[x][y], { reveal: true });
      this.setState({
        board,
        firstClick: false,
        minePositions
      });
      this.interval = setInterval(this.tick, 1000);
    } else {
      this.revealCell(x, y);
    }
  }

  handleContextClickCell(x, y) {
    const { gameOver, won } = this.props;
    const { board } = this.state;
    if (gameOver || won) {
      return;
    }
    if (!board[x][y].reveal) {
      this.toggleFlag(x, y);
    }
  }

  toggleFlag(x, y) {
    const board = this.state.board;
    const { flagged } = board[x][y];
    board[x][y] = Object.assign({}, board[x][y], { flagged: !flagged });
    this.setState({ board });
    this.props.dispatch(toggle(!flagged));
  }

  changeGameLevel(e) {
    const level = e.target.dataset.txt;
    this.props.dispatch(gameLevel(level));
    this.setState({
      board: createBoard(level),
      firstClick: true,
      seconds: 0
    });
    clearInterval(this.interval);
  }

  revealCell(x, y) {
    const board = this.state.board;
    let { level, dispatch } = this.props;
    const { boardWidth, boardHeight } = setGameLevel[level];

    if (!board[x][y].reveal) {
      let mineCount = revealNeighMineCount(board, boardWidth, boardHeight, x, y);
      board[x][y] = Object.assign({}, board[x][y], { reveal: true, mineCount: mineCount })
      this.setState({ board });

      if (mineCount === 0 && !board[x][y].mine) {
        for (let i = x - 1; i <= x + 1; i++) {
          for (let j = y - 1; j <= y + 1; j++) {
            if ((i < 0 || i >= boardWidth) || (j < 0 || j >= boardHeight) || (i === x && j === y) || (board[i][j].flagged)) {
              continue;
            }
            this.revealCell(i, j);
          }
        }
      }

      if (board[x][y].flagged) {
        this.toggleFlag(x, y);
      }

      if (board[x][y].mine) {
        dispatch(gameOver());
        this.state.minePositions.forEach(place => {
          board[place.x][place.y] = Object.assign({}, board[place.x][place.y], { reveal: true });
          this.setState({ board });
        });
        clearInterval(this.interval);
      }

      if (this.hasWon(board)) {
        dispatch(won());
        clearInterval(this.interval);
      }
    }
  }

  hasWon(board) {
    let revealCount = 0;
    const { level } = this.props;
    const { boardWidth, boardHeight, totalMines } = setGameLevel[level];

    board.forEach((row) => {
      row.forEach((cell) => {
        if (cell.reveal && (cell.mine === cell.flagged)) {
          revealCount++;
        }
      });
    });

    return revealCount === (boardWidth * boardHeight) - totalMines;
  }

  render() {
    const { board, seconds } = this.state;
    const { gameOver, won, mine } = this.props;

    return (
      <div className="container">
        <Title />
        <GameLevel
          changeGameLevel={this.changeGameLevel}
        />
        <GameStatus
          gameOver={gameOver}
          seconds={seconds}
          won={won}
          mine={mine}
          reset={this.reset}
        />
        <Board
          board={board}
          onClick={this.handleClickCell}
          onContextClick={this.handleContextClickCell}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => state.gameReducer;

export default connect(mapStateToProps)(App);
