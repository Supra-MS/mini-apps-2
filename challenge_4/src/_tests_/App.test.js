import { render, screen, fireEvent } from '@testing-library/react';
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import { Provider } from 'react-redux';
import { applyMiddleware, createStore } from 'redux';
import reducers from '../reducers';
import gameReducer from '../reducers/gameReducer';
import App from '../components/App';
import GameStatus from '../components/GameStatus';
import { plantMines } from '../utils/setBoard';
import * as actions from '../actions';
import { winBoard } from './data';
import renderer from 'react-test-renderer';

configure({ adapter: new Adapter() });
const store = createStore(reducers, applyMiddleware());
const wrapper = shallow(<App store={store} />);
const instance = wrapper.dive().instance();
let interState, clickBoard, clickMinePositions;

beforeEach(() => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
});

describe('Minesweeper Game Tests:', () => {
  test('Check title', () => {
    const title = screen.getByText(/minesweeper/i);
    expect(title).toBeInTheDocument();
  });

  test('Check the total count for planting mines based on the game level', () => {
    const beginnerCount = plantMines('beginner');
    const intermediateCount = plantMines('intermediate');
    const expertCount = plantMines('expert');

    expect(beginnerCount.length).toBe(10);
    expect(intermediateCount.length).toBe(40);
    expect(expertCount.length).toBe(99);
  });

  test('Check that the First click should not be a mine', () => {
    jest.spyOn(instance, 'handleClickCell');
    expect(instance.state.firstClick).toBe(true);
    expect(instance.state.minePositions).toBeFalsy();
    expect(instance.interval).toBeFalsy();

    instance.handleClickCell(3,1);
    clickBoard = instance.state.board;
    clickMinePositions = instance.state.minePositions;

    expect(instance.state.firstClick).toBe(false);
    expect(clickMinePositions).toBeTruthy();
    expect(instance.interval).toBeTruthy();
    expect(clickBoard[3][1].mine).toBe(false);
  });

  test('Check that on clicking mine dispatches game over()', () => {
    let x = clickMinePositions[0].x;
    let y = clickMinePositions[0].y;

    jest.spyOn(instance, 'revealCell');
    expect(clickBoard[x][y].mine).toBe(true);
    expect(clickBoard[x][y].reveal).toBe(false);
    expect(clickBoard[x][y].flagged).toBe(false);

    instance.revealCell(x, y);
    expect(clickBoard[x][y].mine).toBe(true);
    expect(clickBoard[x][y].reveal).toBe(true);
    expect(clickBoard[x][y].flagged).toBe(false);
    expect(screen.getByText(/ðŸ˜£/i)).toBeInTheDocument();
    fireEvent.click(screen.getByText('ðŸ˜£'));
  });

  test('Check the mine count through toggling flag and reset methods', () => {
    expect(wrapper.exists()).toBe(true);

    jest.spyOn(instance, 'toggleFlag');
    instance.toggleFlag(1, 2);
    expect(instance.toggleFlag).toHaveBeenCalledTimes(1);
    instance.toggleFlag(1, 3);
    expect(instance.toggleFlag).toHaveBeenCalledTimes(2);
    expect(instance.props.store.getState().gameReducer.mine).toBe(8);

    // Clicking same cell
    instance.toggleFlag(0, 3);
    expect(instance.toggleFlag).toHaveBeenCalledTimes(3);
    expect(instance.props.store.getState().gameReducer.mine).toBe(7);
    instance.toggleFlag(0, 3);
    expect(instance.toggleFlag).toHaveBeenCalledTimes(4);
    expect(instance.props.store.getState().gameReducer.mine).toBe(8);

    // After reset
    fireEvent.click(screen.getByText('ðŸ™‚'));
    expect(instance.props.store.getState().gameReducer.mine).toBe(10);
    expect(instance.props.store.getState().gameReducer.level).toEqual('beginner');
    expect(instance.props.store.getState().gameReducer.gameOver).toBe(false);
    expect(instance.props.store.getState().gameReducer.won).toBe(false);
  });


  test('Check the action and reducer for initial start', () => {
    const prevState = { level: "intermediate", mine: 0, gameOver: true, won: true, };
    const actionStart = { type: 'START', };

    const newState = gameReducer(prevState, actionStart);
    expect(newState.level).toEqual('intermediate');
    expect(newState.mine).toEqual(40);
    expect(newState.gameOver).toBe(false);
  });

  test('Check the state of different game levels through reducer and events', () => {
    const prevState = { level: "expert", mine: 0, gameOver: true, won: true, };
    const actionLevel = { type: 'GAME_LEVEL', payload: { level: 'beginner' } };

    const newState = gameReducer(prevState, actionLevel);
    expect(newState.level).toEqual('beginner');
    expect(newState.mine).toEqual(10);
    expect(newState.gameOver).toBe(false);
    fireEvent.click(screen.getByText(/beginner/i));
    expect(instance.props.store.getState().gameReducer.level).toEqual('beginner');
    fireEvent.click(screen.getByText(/expert/i));
    expect(instance.props.store.getState().gameReducer.level).toEqual('expert');
    fireEvent.click(screen.getByText(/intermediate/i));
    interState = instance.props.store.getState().gameReducer;
    expect(instance.props.store.getState().gameReducer.level).toEqual('intermediate');
    expect(instance.props.store.getState().gameReducer.gameOver).toBe(false);
  });

  test('Check the winner status by changing the state and passing through dispatch()', () => {
    jest.spyOn(instance, 'hasWon');
    expect(instance.hasWon(winBoard)).toBe(true);

    const newState = gameReducer(interState, instance.props.store.dispatch(actions.won()));
    expect(newState.level).toEqual('intermediate');
    expect(newState.gameOver).toBe(false);
    expect(newState.won).toBe(true);

    let props = { gameOver: false, won: true, };

    let gameWrapper = renderer.create(<GameStatus {...props} />).toJSON();
    expect(gameWrapper).toMatchSnapshot();
    expect(gameWrapper.children[1].props).toEqual({"className": "win-status won false"});
    expect(gameWrapper.children[1].children[0]).toEqual('ðŸ¥³ CONGRATS! YOU WON ðŸ¥³');
    expect(screen.getByText(/ðŸ˜Ž/i)).toBeInTheDocument();

    props = { gameOver: true, won: false };

    gameWrapper = renderer.create(<GameStatus {...props} />).toJSON();
    expect(gameWrapper.children[1].props).toEqual({"className": "win-status false lose"});
    expect(gameWrapper.children[1].children[0]).toEqual('!! TRY AGAIN !!');
  });

});