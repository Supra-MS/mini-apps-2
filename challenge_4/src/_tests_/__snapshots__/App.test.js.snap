// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Minesweeper Game Tests: Check the winner status by changing the state and passing through dispatch() 1`] = `
<div
  className="container"
>
  <div
    className="game-status"
  >
    <button
      className="time-counter"
    >
      ⏱ 
      undefined
       
    </button>
    <button
      className="reset"
    >
       
      😎
       
    </button>
    <button
      className="mine-counter"
    >
      💣 
       
    </button>
  </div>
  <div
    className="win-status won false"
  >
    🥳 CONGRATS! YOU WON 🥳
  </div>
</div>
`;
